### PR TITLE
Error if reazon-defrel is run under dynamic scope

### PR DESCRIPTION
**Pull request by ethan-leba**
explicitly alerts a user if their relation won't be defined properly

thanks for this cool project!

### CODE REVIEW

**Comment by nickdrozd** [comment]
Congratulations, you are the first external code contributor to Reazon! The code looks good, but I don't have time to test it just now. Hopefully I'll have some time within the next week.

It looks like you ran into the dynamic scope problem. Aside from that, did you have any trouble running anything? Did you run anything from _The Reasoned Schemer_?

**Comment by ethan-leba** [comment]
Sounds good! As far as I've used reazon, it works perfectly fine -- i just tried to run some code in a scratch buffer and had about 30s of confusion why the relation wasn't working

File: reazon.el [hunk 1]
```diff
@@ -425,10 +425,12 @@ Also known as committed choice. This operator is impure."
...
-    `(defun ,name ,varlist
-       (lambda (,stream)
-         (lambda ()
-           (funcall (reazon-conj ,@goals) ,stream))))))
+    `(if (not lexical-binding)
+         (user-error "Can't define new relations when lexical binding is disabled!")
+       (defun ,name ,varlist
+         (lambda (,stream)
+           (lambda ()
+             (funcall (reazon-conj ,@goals) ,stream)))))))
```
Comment (ethan-leba) [inline reazon.el:433]: moved the lexical check into the macro expansion. for some reason the byte-compiler really doesn't like my change:

```lang
make compile
emacs -Q --batch --directory ./ --funcall batch-byte-compile .//reazon-perf.el .//reazon-sudoku.el .//reazon.el
emacs -Q --batch --directory ./ --directory ./test --funcall batch-byte-compile ./test/reazon-test-interface.el ./test/reazon-test-prolog.el ./test/reazon-test-puzzle.el ./test/reazon-test-relation.el ./test/reazon-test-sudoku.el ./test/reazon-test-trs.el ./test/reazon-test-unit.el ./test/reazon-test-utils.el

In toplevel form:
test/reazon-test-interface.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-prolog.el:27:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-puzzle.el:27:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-relation.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-sudoku.el:28:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-trs.el:29:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-unit.el:30:1:Error: Can’t define new relations when lexical binding is disabled!

In toplevel form:
test/reazon-test-utils.el:31:1:Error: Can’t define new relations when lexical binding is disabled!
make: *** [compile] Error 1
```

i really have no clue why this is happening!

Comment (nickdrozd) [reply reazon.el:433]: Somehow these definitions are getting evaluated outside of lexical scoping. But where? `reazon.el` has lexical scoping and so does `bytecomp.el`. I don't understand who's in control of the scoping when batch compilation is run.

**Comment by nickdrozd** [comment]
What version of Emacs are you running? Apparently in 27 lexical scoping was made the default in Scratch, IELM, etc, so I have had some trouble reproducing the old bug at all.

**Comment by ethan-leba** [comment]
I'm on 27.2, so it might just be something with doom emacs -- i'm just gonna drop this PR for now since i don't feel like digging into the byte compiler internals x_x

**Pull request closed without merging**